### ISSUE TITLE
transferring large ontology files from their canonical iri can be slow

### ISSUE THREAD

**Issue by tgbugs** [OBOFoundry/purl.obolibrary.org#162]
Importing some large OBO ontologies directly from their canonical purl.obolibrary.org IRIs is extremely slow. When building our ontologies we often want to use the canonical IRI for imports to simplify maintenance, provenance, and synchronization. However, this is practically difficult in our infrastructure when a canonical IRI can take many minutes to transfer an ontology file. In the short term are a couple of solutions that we could employ (@cmungall has one solutions listed [here](https://github.com/owlcollab/owltools/wiki/Import-Chain-Mirroring)), however in the long run many folks would benefit from being able to use canonical IRIs for ontologies in their production systems.

Would it be possible to develop a kind of CDN for ontology files either in a centralized or federated manner so that canonical IRIs could serve ontologies with sufficient bandwidth for production use?

It looks like there is also a related issue for supporting compression at https://github.com/owlcs/owlapi/issues/348.

**Comment by cmungall** [comment]
We should discuss this on the next obo-tech call

also cc @kltm

**Comment by jamesaoverton** [comment]
Currently each OBO project has its own infrastructure. We've discussed this sort of shared infrastructure for distribution. IIRC, the leading contender was Amazon S3. Would S3 fit your needs @tgbugs?

The other half of the story is better caching, using stuff like HTTP ETag to avoid unnecessary downloads.

I'm very interested in pursuing this topic, and making it easy for our community to use via a tool such as [ROBOT](https://github.com/ontodev/robot). ROBOT https://github.com/ontodev/robot/issues/3 is related.

**Comment by tgbugs** [comment]
@jamesaoverton we are currently transitioning to Amazon for a number of things so S3 would probably fit our needs quite well. IIRC @cmungall ran some numbers on S3 and it looked like it might be a bit on the expensive side which made me think it might worth exploring other options (though I wonder how that would change if everything going over the pipe was compressed).

**Comment by cmungall** [comment]
For non-massive ontologies, pushing to github and using the github 
release mechanism is often sufficient, and has various benefits. Even 
for larger ontologies, using a modular approach can help here.

However, it could be useful to mirror ontologies and their versioned 
purls on S3 or similar. I'm imagining something working in a similar way 
to zenodo.

**Comment by cmungall** [comment]
It would be worth doing this again.

Our numbers were based on ontology files served from 
berkeleybop.org/ontologies, which is a subset of the overall file 
content

With the new purl system under our control and not in OCLC, can we get 
some stats about how often different ontology artefacts are downloaded.

Overall numbers, e.g. for class purls as well would be v useful to have

**Comment by kltm** [comment]
For cheap, nothing beats the free infrastructure transfer at an institution. Redundancy and fast DNS control can be supplied by AWS (Route 53, etc.), with the actual file hosting occurring on file servers at a couple of different sites. More complicated than an all AWS solution, but definitely cheaper and likely more robust. A "create mirror" ansible script could be easily setup, allowing people to grab things for local use or quickly create a fallback in case of failure.

**Comment by tgbugs** [comment]
I was running tests on this again and noticed that some ontologies served from the cloudfront CDN do not seem to accept gzip encoding. For example `curl --verbose --header 'Accept-encoding: gzip' --compressed --location --remote-name-all http://purl.obolibrary.org/obo/{go,uberon}.owl`. The compression ratios are ~23x go and ~17x for uberon.

I think that the reason this is happening is because  the original server always returns `Content-Type: application/rdf+xml` which cloudfront compression does not support according to [the docs](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/ServingCompressedFiles.html). I tested by requesting other content types (text/plain, text/xml, application/xml) but I may not have hit one that was valid since the response was alwasy application/rdf+xml.

Any chance additional content types could be supported, or gzipping of the application/rdf+xml by the original host implemented?

**Comment by jamesaoverton** [comment]
@tgbugs: The PURL system in this repository only does redirects. Since it doesn't serve content, it doesn't reply with a `Content-Type`. So (if I understand correctly) the issue you're describing applies to the infrastructure provided by BBOP, @cmungall @kltm.

**Comment by cmungall** [comment]
this is useful to know thanks @tgbugs

**Comment by tgbugs** [comment]
@jamesaoverton yes, that is correct, I added it as a follow up on this issue since the purl system seems ideally situated to make checks as @cmungall suggests.
